### PR TITLE
Pangocairo

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc @EPashkin

It currently needs this [update](https://github.com/gtk-rs/gir-files/pull/10) to work. Then I'll update the `Cargo.toml`.

(I'll clean everything once the other one is merged because the history seems like a huge mess.)

### CODE REVIEW

**Comment by EPashkin** [comment]
@GuillaumeGomez you don't forget about https://crates.io/crates/pangocairo ?

**Comment by GuillaumeGomez** [comment]
It's strangely done and outside of `gtk-rs` (and I didn't find the `Gir.toml` for the non-sys generation, seems a bit weird). Also, I think this one will be the last we'll add in `gtk-rs`.

**Comment by EPashkin** [comment]
We was decided that @RazrFalcon will handle this crate.
It already has same fix https://github.com/RazrFalcon/pangocairo-rs/blob/master/pangocairo-sys/gir-patch.diff
Strange that it don't contains `Gir.toml` or build script.

**Comment by EPashkin** [comment]
As I remember main reason for not including `pangocairo` to `gtk-rs` was patched .gir.
See https://github.com/gtk-rs/gtk/issues/453

**Comment by RazrFalcon** [comment]
I still have no idea how gir works, but crate itself if working well. I don't know what can be changed there.

**Comment by GuillaumeGomez** [comment]
@RazrFalcon: It's mainly in order to be able to easily update by just running `make gir`. Also, I think it's the last piece missing and since it's not the first time someone is asking for it, we should integrate it into `gtk-rs`.

@EPashkin: Is it really an issue to have some `.gir` files patched?

**Comment by EPashkin** [comment]
For me, it almost not problem is patch applied on same script run that do download.

**Comment by GuillaumeGomez** [comment]
Great, I can do that. Just las remaining question: @RazrFalcon do you want we get your repository and clean it up so your work is kept or should I just go over with this repository?

**Comment by RazrFalcon** [comment]
@GuillaumeGomez I don't really care. I can remove this repos if there is official one. But what to do with crates itself? Can I transfer an ownership to you?

**Comment by GuillaumeGomez** [comment]
@RazrFalcon Yes you can (and thanks a lot!): http://doc.crates.io/crates-io.html#cargo-owner

**Comment by RazrFalcon** [comment]
@GuillaumeGomez can you provide a full command?

**Comment by GuillaumeGomez** [comment]
The command is the following:

```bash
cargo owner --add GuillaumeGomez
```

And if you want to get removed from the crate owners (but I don't mind if you remain as the second crate owner):

```bash
# /!\ be sure that you really want to not be the second crate maintainer before launching the command!
cargo owner --remove RazrFalcon
```

**Comment by RazrFalcon** [comment]
@GuillaumeGomez done for `pangocairo` and `pangocairo-sys` (should be removed, AFAIU).

**Comment by GuillaumeGomez** [comment]
@RazrFalcon: Thanks a lot! And no, we need the `-sys` crate as well. :)

**Comment by GuillaumeGomez** [comment]
Ah but you're right, we'll put it into the `sys` repository...

**Comment by GuillaumeGomez** [comment]
@EPashkin: Updated.

File: pango-cairo-sys/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#[allow(unused_imports)]
+use glib::{gboolean, gconstpointer, gpointer, GType, Volatile};
+
+// Callbacks
+pub type PangoCairoShapeRendererFunc = Option<unsafe extern "C" fn(*mut cairo::cairo_t, *mut pango::PangoAttrShape, gboolean, gpointer)>;
+
+// Classes
+//#[repr(C)]
+//pub struct PangoCairoFcFontMap {
+//    pub parent_instance: /*Metadata mismatch*/[c:type mismatch PangoFcFontMap != PangoCairoFcFontMap of FcFontMap],
```
Comment (EPashkin) [inline pango-cairo-sys/src/lib.rs]: Seems gir-files fix not much help :(

Comment (GuillaumeGomez) [reply pango-cairo-sys/src/lib.rs]: That's not the only problem apparently. This gir file is a mess.

File: conf/gir-pango-cairo.toml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[options]
+work_mode = "sys"
+library = "PangoCairo"
+version = "1.0"
+min_cfg_version = "1.30"
+external_libraries = [
+    "GLib",
+    "Pango",
+    "Cairo",
```
Comment (EPashkin) [inline conf/gir-pango-cairo.toml]: Seems this external cause problem as it don't have `cairo-sys` here

Comment (GuillaumeGomez) [reply conf/gir-pango-cairo.toml]: It doesn't make sense? It's only using `sys` crates after all so adding `cairo-sys` in here couldn't work. Or did I miss something?

Comment (EPashkin) [reply conf/gir-pango-cairo.toml]: Try manually edit pangocairo-sys/Cargo.toml for cairo-sys path. Seems it will be not replaced on update

Comment (GuillaumeGomez) [reply conf/gir-pango-cairo.toml]: I did, it's not a problem anymore. ;)

Now remains two problems:

 1. FcFontMap (which seems correct to me as defined in the gir file...).
 2. cairo::FontType (it's an enum in cairo and gir doesn't seem to understand when I say that).

Comment (GuillaumeGomez) [reply conf/gir-pango-cairo.toml]: Ok, number 2 is solved in a... veeeeeeery ugly way. But whatever, it works haha.

Comment (EPashkin) [reply conf/gir-pango-cairo.toml]: Seems this type missed from pango.gir https://developer.gnome.org/pango/stable/PangoFcFontMap.html, so this gir looks as class is parent of yourself.
```
    <class name="FcFontMap"
           c:symbol-prefix="fc_font_map"
           c:type="PangoCairoFcFontMap"
           parent="Pango.FontMap"
           glib:type-name="PangoCairoFcFontMap"
           glib:get-type="pango_cairo_fc_font_map_get_type">
      <implements name="FontMap"/>
      <field name="parent_instance" introspectable="0">
        <type name="FcFontMap" c:type="PangoFcFontMap"/>
      </field>
```

Comment (GuillaumeGomez) [reply conf/gir-pango-cairo.toml]: When I said this gir file was a huge mess, I wasn't kidding you know? :p

**Comment by GuillaumeGomez** [comment]
And updated again.

**Comment by EPashkin** [comment]
At minimum it compiles now, so 👍

**Comment by GuillaumeGomez** [comment]
Yeay!

**Pull request merged**